src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -983,6 +1003,13 @@ ucs_status_t uct_dc_mlx5_ep_fc_ctrl(uct_ep_t *tl_ep, unsigned op,
     }
 
     return UCS_OK;
+
+out_dci_put:
+    if (init_dci_index == UCT_DC_MLX5_EP_NO_DCI) {
```

Comment:
why need to check initial index? Wouldn't it be checked in `uct_dc_mlx5_iface_dci_put`?

Reply:
> why need to check initial index?

we have to check it to understand whether we need to release DCI here or not. there were DCI leak here before my change.
added a comment.

> Wouldn't it be checked in uct_dc_mlx5_iface_dci_put?

we have to check initial DCI - not the current one. The current one is always set here or previously.

Reply:
what is the difference between initial and current? It is either `NO_DCI` or some specific dci.

Reply:
so, can we avoid checking `init_dci_index`?

Reply:
> what is the difference between initial and current?

- initial DCI - the value in `ep->dci` stored when we entered the function (it could be `NO_DCI`)
- current DCI = the value in `ep->dci` stored when we reached `out_dci_put` label (it could be only some specific DCI)

So, if initial DCI is `NO_DCI`, we should release DCI stored stored in `ep->dci`. Otherwise, we should not release.

Reply:
if you have something outstanding, dci will not be released, otherwise it will be released which is what you need.
My point is that it does not seem to be necessary to check for initial dci. You can just put it

Reply:
yes,
but if we don't have outstanding operations, e.g. timeout for re-sending FC_HARD_REQ isn't expired yet, we have to release DCI in case of it wasn't allocated before doing FC_HARD_REQ.
otherwise, we will release DCI, but there are outstanding operations